cupy/core/_reduction.pyx

```diff
@@ -290,13 +299,118 @@ cdef class _AbstractReductionKernel:
+            block_size, block_stride, out_block_num = _get_block_specs(
+                internal.prod(in_shape),
+                internal.prod(out_shape),
+                contiguous_size, -1)
+        else:
+            # Optimize dynamically
+
+            # Calculate a key unique to the reduction setting.
+            for x in in_args + out_args:
+                if isinstance(x, ndarray):
+                    shape_and_strides.append(x.shape)
+                    shape_and_strides.append(x.strides)
+                else:
+                    shape_and_strides.append(None)
+                    shape_and_strides.append(None)
+            key = (
+                id(self), tuple(shape_and_strides),
+                in_types, out_types, reduce_type, device_id,
+            )
+
+            params = optimize_context.get_params(key)
+            if params is None:
+                params = self._get_optimized_params(
+                    optimize_context.config, in_args, out_args,
+                    in_shape, out_shape, type_map, map_expr, reduce_expr,
+                    post_map_expr, reduce_type, stream)
+                optimize_context.set_params(key, params)
+            block_size, block_stride, out_block_num = params
+
+        # Launch the kernel
+        self._launch(
+            out_block_num,
+            block_size,
+            block_stride,
+            in_args, out_args,
+            in_shape, out_shape,
+            type_map,
+            map_expr, reduce_expr, post_map_expr, reduce_type,
+            stream)
+        return ret
+
+    def _get_optimized_params(
+            self, optimize_config, in_args, out_args, in_shape, out_shape,
+            type_map, map_expr, reduce_expr, post_map_expr, reduce_type,
+            stream):
+        out_size = internal.prod_sequence(out_shape)
+
+        def copy_arg(a):
+            if isinstance(a, ndarray):
+                x = _create_ndarray_from_shape_strides(
+                    a._shape, a._strides, a.dtype)
+                assert a.data.device_id == x.data.device_id
+                x[:] = a
+                return x
+            return a
+
+        in_args = [copy_arg(a) for a in in_args]
+        out_args = [copy_arg(a) for a in out_args]
+
         contiguous_size = _get_contiguous_size(
-            in_args, self.in_params, in_shape.size(), out_shape.size())
-        block_size, block_stride, out_block_num = _get_block_specs(
+            in_args, self.in_params, len(in_shape), len(out_shape))
+        block_size, block_stride, default_out_block_num = _get_block_specs(
             internal.prod(in_shape),
             internal.prod(out_shape),
-            contiguous_size)
+            contiguous_size, -1)
+        default_block_size_log = math.floor(math.log2(block_size))
+        default_block_stride_log = math.floor(math.log2(block_size))
+
+        def target_func(block_size, block_stride, out_block_num):
+            self._launch(
```

Comment:
Is it possible to remove `target_func` and `suggest_func` from `_reduction.pyx`.
This is Optuna specific code so having it self-contained will be great for maintainability.

If its difficult or impacts performance lets leave it as it is.

Reply:
`optimize_impl` taken from `optimize_config` in `_reduction.pyx` is always `_optimize` defined in `_optimize.py`, so you may feel it complicated certainly, but this design is intended. There are mainly two reasons for the current design.
1. Some other features (e.g., custom kernel fusion) may use parameter optimization, so generic parts should be split out from `_reduction.pyx`.
2. `optuna` should not be imported when the users use reduction routines without parameter optimization. So, I don't want to import `optuna` nor `_optimize.py` in `_reduction.pyx`.

Reply:
Continuing this discussion, wouldn't elementwise kernels and ufuncs also benefit from this optimization? Sounds like you'd need to duplicate some code in `cupy/core/_kernel.pyx`? 

On the import of `optuna`, I suppose it can be deferred until one enters the context manager `with optimizing.optimize(...)`.